### PR TITLE
Fix hybrid MTP partial CUDA graph capture

### PR DESCRIPTION
**Pull request by guihong-nv**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Fix Transformer Engine partial CUDA-graph capture for hybrid mHC models by including hybrid MTP layers and grouping adjacent attention and MoE router/preprocessing work into one logical graph.

The hybrid stack represents attention and MoE as separate physical layers, while the GPT path captures both scopes in one callable. Hybrid MTP also wraps its internal graphable layers in a non-graphable `HybridStack`, so current discovery omits MTP entirely. This change:

- descends into hybrid MTP stacks during TE callable discovery;
- groups supported adjacent mHC attention and partial-MoE layers;
- keeps expert communication/compute and BDA eager;
- preserves module registration and checkpoint keys while exposing graph-covered tail parameters to TE's autograd input surface;
- propagates the MTP microbatch ID to flattened inner layers.

Grouping is deliberately disabled for full recompute and mHC-selective recompute, whose scheduling differs from the normal hybrid loop.

## Issue tracking

Linked issue: N/A (small bug fix)

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

## Validation

- Targeted unit tests: 2 passed.
- Formatting/static checks: `black --check`, `ruff check`, `isort --check-only`, `py_compile`, and `git diff --check` passed.
- One-node Nsight validation: three combined forward graphs (two decoder + one MTP); MLA and router/top-k signatures are fully graphed with zero eager instances.
- Four-node/16-GPU EP16 training completed successfully with MTP and partial CUDA graphs; no skipped or NaN iterations.
- Same-node-set EP16 ABBA comparison (two baseline and two fixed 32-iteration runs):
  - 12.30% higher normal steady-state throughput;
  - 8.10% higher throughput including periodic manual-GC pauses;
  - all 32 logged LM/MTP/load-balancing losses and gradient norms matched exactly at logger precision across all four runs.

The runtime validation uses a reduced BF16 DSv4-hybrid reproduction; full-scale MXFP8 performance is not claimed.

### CODE REVIEW

**Comment by guihong-nv** [comment]
/ok to test [1b420d6](https://github.com/NVIDIA/Megatron-LM/pull/6583/commits/1b420d67e9123da051f3c754ffc69ea13779e666)

File: megatron/core/models/hybrid/hybrid_block.py [hunk 1]
```diff
@@ -984,6 +1071,11 @@ def get_inner_quant_context(config, layer_number):
                                 packed_seq_params=packed_seq_params,
                             )
 
+                    if isinstance(layer, HyperConnectionHybridLayer) and getattr(
```
Comment (FDecaYed) [inline megatron/core/models/hybrid/hybrid_block.py]: Could you take another look at the eval path here? `GraphableMegatronModule` only uses TE graphs while `self.training` is true, but this skip is based solely on `cuda_graphs` being populated. After graphs have been created, validation switches the model to eval without deleting them, so the attention head runs eager while the grouped MoE tail appears to be skipped. Is that the intended behavior?

File: megatron/core/transformer/cuda_graphs.py [hunk 2]
```diff
@@ -1701,6 +1701,55 @@ def _layer_is_graphable(layer, config):
...
+
+        if layer_number + 1 < len(layers):
+            next_layer = layers[layer_number + 1]
+            can_group = getattr(layer, '_can_group_te_cuda_graph_with', None)
+            if (
+                can_group is not None
+                and _layer_is_graphable(next_layer, config)
+                and can_group(next_layer)
+            ):
+                layer._set_te_cuda_graph_group_tail(next_layer)
```
Comment (FDecaYed) [inline megatron/core/transformer/cuda_graphs.py:1722]: Could you check how this grouping interacts with `--first-last-layers-bf16`? The tail is removed from the callable list here, while `fp8_enabled` is later computed from each retained callable’s `layer_number`. If an attention/MoE pair straddles a BF16 boundary, would the tail inherit the head’s FP8/BF16 setting instead of its own layer setting?

**Review by FDecaYed — CHANGES_REQUESTED** [review]
Please take a look at the two inline questions.

**Comment by guihong-nv** [comment]
/ok to test [014ac0c](https://github.com/NVIDIA/Megatron-LM/pull/6583/commits/014ac0cefabbcb9110b634dde832bb14c32accf1)

**Review by FDecaYed — APPROVED** [review]
Both requested cases are addressed on 014ac0ce: eval no longer skips the grouped tail, and grouping no longer crosses first/last BF16 boundaries. Exact-head CI is green. Looks good to me.
